src/FastRouteRouter.php

```diff
@@ -277,41 +261,50 @@ public function generateUri($name, array $substitutions = [], array $options = [
         }
 
         $route   = $this->routes[$name];
-        $path    = $route->getPath();
         $options = ArrayUtils::merge($route->getOptions(), $options);
 
         if (! empty($options['defaults'])) {
             $substitutions = array_merge($options['defaults'], $substitutions);
         }
 
-        foreach ($substitutions as $key => $value) {
-            $pattern = sprintf(
-                '~%s~x',
-                sprintf(self::VARIABLE_REGEX, preg_quote($key))
-            );
-            $path = preg_replace($pattern, $value, $path);
-        }
+        $routeParser = new RouteParser();
+        $routes = $routeParser->parse($route->getPath());
+
+        // One route pattern can correspond to multiple routes if it has optional parts
+        foreach ($routes as $r) {
+            $path = '';
+            $paramIdx = 0;
+            foreach ($r as $part) {
+                // Fixed segment in the route
+                if (is_string($part)) {
+                    $path .= $part;
+                    continue;
+                }
+
+                // Placeholder in the route
+                if ($paramIdx === count($substitutions)) {
+                    throw new LogicException('Not enough parameters given');
```

Comment:
Clearly I rushed through the review a bit hastily!

Yes, agreed with @michaelmoussa — it's far more helpful to the developer to know what was expected, and what was actually provided, as they can then figure out what they missed in their code.